Katello/katello
# Fixes #35296 - New Custom CDN type

Now that we have the syncable yum exports available, we need the ability to pull in the exported content.
One should be able to enable via http / https to a custom cdn location.

However while enabling a repo via https we hit an SSL verifcation issue. This is because we only pass redhat ca cert while talking to cdn. So the connection is not able to verify the custom cdn url.
This commit adds a new 'custom_cdn'  type configuration where you specify both your custom cdn url and ssl cert.
 
#### What are the changes introduced in this pull request?
- New Custom CDN configuration type. API + UI Changes
- Updated the RedHat CDN type and set its url as read only.

#### Considerations taken when implementing this change?
Initially thought of just using the default trust store to accept custom cert. However from a security perspective we probably shouldn't be making use of those.

#### What are the testing steps for this pull request?
Feel free to ping me for the dump url or do the following 7 steps to generate one.
- Enable a RH Repo like `rhel-7-server-ansible-2.9-rpms`
- Go to products -> repo details of this repo and mark the download policy  immediate
- Sync repo
- On your dev env you may need to run sudo setfacl --recursive --modify u:vagrant:rwX,d:u:vagrant:rwX /var/lib/pulp/exports so that listing files can be generate when hammer is running as a vagrant user.
- `hammer content-export complete repository --id=<repo-id> --format=syncable` -> This should generate a dump
- `mv <dump-dir>  /var/www/html/pub/repos`
- make sure `https://<...>/pub/repos` is browsable.

Now in a new Katello instance which will be the importing sat
- create new org
- import manifest
- Goto Subscriptions => Manage Manifest => CDN Configuration => Redhat CDN
-  Change the URL to https://<...>/pub/repos
- Click updateb
- Now go to Redhat Repos page and try enabling the ansible repo

Before this PR - 
You should see an error like 
```
2022-07-28T18:01:33 [I|app|37e3cb19] CDN: Requesting path https://<webserver>:443/pub/repos/content/dist/rhel/server/7/listing
/opt/rh/rh-ruby27/root/usr/share/ruby/net/protocol.rb:44: warning: exception in verify_callback is ignored
2022-07-28T18:01:33 [E|app|37e3cb19] Failed at scanning for repository: SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate in certificate chain)
```

After PR
- `bundle exec rake db:migrate`
- Go to `Content->Content Credentials`
- Create a new cert with contents of -> `https://<...>/pub/katello-server-ca.crt`
- Go to Subscriptions => Manage Manifest => CDN Configuration => Custom CDN
- Update the url to `https://<...>/pub/repos`  and content credentials to the one you created above
- Now try enabling and syncing the desired repo.
